File: LeetCode/3568. Minimum Moves to Clean the Classroom/solution.cpp
#define MAXN 20
#define MAXL 10
#define MAXE 50
// 这么大的数组开在函数里会爆栈，所以开在全局
int dis[MAXN][MAXN][MAXE + 1][1 << MAXL];

class Solution {
public:
    int minMoves(vector<string>& classroom, int energy) {
        int n = classroom.size(), m = classroom[0].size();
        int L = 0, SI, SJ;
        for (int i = 0; i < n; i++) for (int j = 0; j < m; j++) {
            // 给每个垃圾一个 0 ~ 9 的编号
            if (classroom[i][j] == 'L') classroom[i][j] = L + '0', L++;
            // 找一下起点
            else if (classroom[i][j] == 'S') SI = i, SJ = j;
        }
        
        short dir[4][2] = {0, 1, 1, 0, -1, 0, 0, -1};
        for (int i = 0; i < n; i++) for (int j = 0; j < m; j++)
            for (int k = 0; k <= energy; k++) for (int l = 0; l < (1 << L); l++)
                dis[i][j][k][l] = -1;
        // BFS
        queue<array<int, 4>> q;
        q.push({SI, SJ, energy, 0}); dis[SI][SJ][energy][0] = 0;
        while (!q.empty()) {
            auto arr = q.front(); q.pop();
            int i = arr[0], j = arr[1], w = arr[2], msk = arr[3];
            // 已经捡完所有垃圾，返回答案
            if (msk == (1 << L) - 1) return dis[i][j][w][msk];
            // 枚举下一步走的方向
            if (w > 0) for (int k = 0; k < 4; k++) {
                int ii = i + dir[k][0], jj = j + dir[k][1];
                if (ii < 0 || jj < 0 || ii >= n || jj >= m || classroom[ii][jj] == 'X') continue;
                int ww = w - 1, mmsk = msk;
                // 走到 R，恢复所有能量
                if (classroom[ii][jj] == 'R') ww = energy;
                // 走到垃圾，记录在 mask 里
                else if (classroom[ii][jj] >= '0' && classroom[ii][jj] <= '9') mmsk |= 1 << (classroom[ii][jj] - '0');
                if (dis[ii][jj][ww][mmsk] >= 0) continue;
                q.push({ii, jj, ww, mmsk}); dis[ii][jj][ww][mmsk] = dis[i][j][w][msk] + 1;
            }
        }
        // 捡不完垃圾
        return -1;
    }
};

作者：TsReaper
链接：https://leetcode.cn/problems/minimum-moves-to-clean-the-classroom/solutions/3690738/bfs-by-tsreaper-d8x2/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: LeetCode/3568. Minimum Moves to Clean the Classroom/solution.py
class Solution:
    def minMoves(self, classroom: List[str], energy: int) -> int:
        m, n = len(classroom), len(classroom[0])
        idx = [[0] * n for _ in range(m)]
        cnt_l = sx = sy = 0
        for i, row in enumerate(classroom):
            for j, b in enumerate(row):
                if b == 'L':
                    idx[i][j] = 1 << cnt_l  # 给垃圾分配编号（提前计算左移）
                    cnt_l += 1
                elif b == 'S':
                    sx, sy = i, j
        if cnt_l == 0:
            return 0

        DIRS = (-1, 0), (1, 0), (0, -1), (0, 1)
        vis = [[[[False] * (1 << cnt_l) for _ in range(energy + 1)] for _ in range(n)] for _ in range(m)]
        vis[sx][sy][energy][0] = True
        q = [(sx, sy, energy, 0)]

        full_mask = (1 << cnt_l) - 1
        ans = 0
        while q:
            tmp = q
            q = []
            for x, y, e, mask in tmp:
                if mask == full_mask:  # 所有垃圾收集完毕
                    return ans
                if e == 0:  # 没能量了
                    continue
                for dx, dy in DIRS:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < m and 0 <= ny < n and classroom[nx][ny] != 'X':
                        new_e = energy if classroom[nx][ny] == 'R' else e - 1
                        new_mask = mask | idx[nx][ny]  # 添加垃圾（没有垃圾时 mask 不变）
                        if not vis[nx][ny][new_e][new_mask]:
                            vis[nx][ny][new_e][new_mask] = True
                            q.append((nx, ny, new_e, new_mask))
            ans += 1
        return -1

作者：灵茶山艾府
链接：https://leetcode.cn/problems/minimum-moves-to-clean-the-classroom/solutions/3690747/bfs-by-endlesscheng-rpk6/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。